docs/key-issues/gpus.md

```diff
@@ -180,3 +204,84 @@ Potential solutions on the PyPI side include:
   declare a dependency on a CUDA or ROCm version as externally provided,
+
+
+## Additional Notes on CUDA Compatibility
+
+Compatibility across CUDA versions is a common problem with numerous pitfalls to be aware of.
+There are three primary components of CUDA:
+
+1. The CUDA Toolkit (CTK): This component includes the CUDA runtime library
+   (libcudart.so) along with a range of other libraries and tools including
+   math libraries like cuBlas. libcudart.so and a few other core headers and
+   libraries compose the bare minimum required to compile CUDA code.
+2. The user-mode driver: This is the libcuda.so library. This library is
+   required to actually run CUDA code.
+3. The kernel-mode driver: The nvidia.ko file. This constitutes what is
+   typically considered a "driver" in common parlance when referring to other
+   peripherals connected to a computer.
+
+CUDA developers and users generally do not need to be aware of the kernel-mode driver.
+For the rest of this section, when referring to the "driver" we will always be referring to the user-mode driver libcuda.so.
+
+The CUDA runtime library makes no forward or backwards compatibility
+guarantees, meaning that libraries that dynamically link to the CUDA runtime
+may not work correctly if they are run on a system with a different CUDA
+runtime shared library than the one they were compiled against. In this scenario,
+users are responsible for having the right CUDA runtime library installed.  As
+a result, the official CUDA recommendation is to statically link the CUDA
+runtime (see
+[here](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#recommendations-for-building-a-minor-version-compatible-library)
+and
+[here](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#distributing-the-cuda-runtime-and-libraries)
+for more information).
+
+CUDA drivers have always been backwards compatible. Any code that runs when
+some driver version X is installed will always work correctly when run with
+some newer driver version Y>X. As briefly discussed above, as of CUDA 11.0
+CUDA also promises [minor version
+compatibility](https://docs.nvidia.com/deploy/cuda-compatibility/index.html#minor-version-compatibility)
+(MVC). This compatibility guarantees that CUDA code compiled using a certain
+version of the CUDA toolkit will be compatible with any driver version within
+the same major release. This behavior is useful because it is often easier for
+users to upgrade their CUDA runtime than it is to upgrade the driver,
+especially on shared machines. For instance, the CUDA toolkit may be installed
+using conda, while the driver library cannot be. An example of leveraging MVC
+would be compiling code against the CUDA 11.5 runtime library and then running
+it on a system with a CUDA 11.2 driver installed. However there are some
+caveats with MVC:
+
+- If the code uses any features that were introduced in a later driver version
```

Comment:
Is `cuGetProcAddress`, introduced in CUDA 11.3, an exception to this? If I attempt to call it using `ctypes` in Python or `dlsym` in C, will I just get a normal linker error?

Reply:
Now I think about it, is this a feature of the runtime, rather than a guarantee made for code that directly uses the driver API?

Reply:
Yeah I was under the impression that MVC only applies to the runtime API, not the driver API. We could test that though to be sure.

Reply:
Minor ver compat applies to the UMD too.

Reply:
Looking over the docs, I can't see anything that specifically talks about direct use of driver APIs - the closest thing I can find is in https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#handling-new-cuda-features-and-driver-apis, which says:

> To use other CUDA APIs introduced in a minor release (that require a new driver), one would have to implement fallbacks or fail gracefully.

@leofang Have you seen anything that refers more specifically to driver APIs?

For my example of `cuGetProcAddress`, I've seen it suggested (elsewhere) that one should fall back to a platform-specific equivalent - e.g. `GetProcAddress()` on Windows, or `dlsym()` on Linux. Maybe we need to note here that direct use of driver APIs added in minor releases needs to be considered / guarded against appropriately, without being too prescriptive (since the thing to do could vary depending on what the API is and the application features / requirements).

Reply:
Yes, we need fallbacks for new APIs introduced in the middle of a major version. But this is nothing unusual. MVC is basically reaffirming that CUDA is obeying semantic versioning IMHO, and it's OK to make non-breaking changes (ex: add new APIs) in a minor release, so I am not sure we should expand on this.

Reply:
If a new API is introduced in CUDA 12.2, then any code using it will need fallbacks/failure modes handling it when compiling on CUDA 12.0/12.1. That's true whether it's a runtime or driver API, which I think is also what Leo is saying.

I'm not really sure how minor version compatibility makes sense for the UMD. The driver APIs are implemented in the same library that provides the UMD (libcuda.so). How would you have a "newer" driver API installed than the UMD?

The whole point of MVC IIUC is that using code compiled against a specific libcudart won't automatically give linker errors if that libcudart finds an older libcuda on the system than the one it was compiled against.